src/nvim/tui/tui.c

```diff
@@ -1318,46 +1308,56 @@ void tui_default_colors_set(TUIData *tui, Integer rgb_fg, Integer rgb_bg, Intege
   invalidate(tui, 0, tui->grid.height, 0, tui->grid.width);
 }
 
-static void tui_set_dirty(TUIData *tui)
-  FUNC_ATTR_NONNULL_ALL
-{
-  if (!tui->dirty) {
-    tui->dirty = true;
-    tui_flush_start(tui);
-  }
-}
-
 /// Begin flushing the TUI. If 'termsync' is set and the terminal supports synchronized updates,
 /// begin a synchronized update. Otherwise, hide the cursor to avoid cursor jumping.
-static void tui_flush_start(TUIData *tui)
+static size_t tui_flush_start(TUIData *tui, char *buf, size_t len)
   FUNC_ATTR_NONNULL_ALL
 {
+  const char *str = NULL;
+
   if (tui->sync_output && tui->unibi_ext.sync != -1) {
     UNIBI_SET_NUM_VAR(tui->params[0], 1);
-    unibi_out_ext(tui, tui->unibi_ext.sync);
+    str = unibi_get_ext_str(tui->ut, (size_t)tui->unibi_ext.sync);
   } else if (!tui->is_invisible) {
-    unibi_out(tui, unibi_cursor_invisible);
+    str = unibi_get_str(tui->ut, unibi_cursor_invisible);
     tui->is_invisible = true;
   }
+
+  if (!str) {
+    return 0;
+  }
+
+  return unibi_run(str, tui->params, buf, len);
 }
 
 /// Finish flushing the TUI. If 'termsync' is set and the terminal supports synchronized updates,
 /// end a synchronized update. Otherwise, make the cursor visible again.
-static void tui_flush_end(TUIData *tui)
+static size_t tui_flush_end(TUIData *tui, char *buf, size_t len)
   FUNC_ATTR_NONNULL_ALL
 {
+  size_t offset = 0;
   if (tui->sync_output && tui->unibi_ext.sync != -1) {
     UNIBI_SET_NUM_VAR(tui->params[0], 0);
-    unibi_out_ext(tui, tui->unibi_ext.sync);
+    const char *str = unibi_get_ext_str(tui->ut, (size_t)tui->unibi_ext.sync);
+    offset += unibi_run(str, tui->params, buf, len);
```

Comment:
Actually using `+=` doesn't make sense here, because it not added to `buf` or subtracted from `len`.